objects/src/accounts/package/storage_entry/mod.rs

```diff
@@ -0,0 +1,611 @@
+
+#[derive(Default, Serialize, Deserialize)]
+/// Used as a helper for validating and (de)serializing storage entries
+struct RawStorageEntry {
+    name: String,
+    description: Option<String>,
+    slot: Option<u8>,
+    slots: Option<Vec<u8>>,
+    value: Option<WordRepresentation>,
+    values: Option<StorageValues>,
+}
+
+impl From<StorageEntry> for RawStorageEntry {
+    fn from(entry: StorageEntry) -> Self {
+        match entry {
+            StorageEntry::Value { name, description, slot, value } => RawStorageEntry {
+                name,
+                description,
+                slot: Some(slot),
+                value: Some(value),
+                ..Default::default()
+            },
+            StorageEntry::Map { name, description, slot, values } => RawStorageEntry {
+                name,
+                description,
+                slot: Some(slot),
+                values: Some(StorageValues::MapEntries(values)),
+                ..Default::default()
+            },
+            StorageEntry::MultiSlot { name, description, slots, values } => RawStorageEntry {
+                name,
+                description,
+                slots: Some(slots),
+                values: Some(StorageValues::Words(values)),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+impl Serialize for StorageEntry {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let raw_storage_entry: RawStorageEntry = self.clone().into();
+        raw_storage_entry.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for StorageEntry {
+    fn deserialize<D>(deserializer: D) -> Result<StorageEntry, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let raw = RawStorageEntry::deserialize(deserializer)?;
+
+        // Determine presence of fields and do early validation
+        let slot_present = raw.slot.is_some();
+        let value_present = raw.value.is_some();
+
+        // Use a match on the combination of presence flags to choose variant
+        match (raw.slots, raw.values) {
+            (None, None) => {
+                // Expect a Value variant: "slot" and "value" must be present
+                // "slots" and "values" must not be present.
+                Ok(StorageEntry::Value {
+                    name: raw.name,
+                    description: raw.description,
+                    slot: raw
+                        .slot
+                        .ok_or(D::Error::custom("Missing 'slot' field for single-slot entry."))?,
+                    value: raw
+                        .value
+                        .ok_or(D::Error::custom("Missing 'value' field for single-slot entry."))?,
+                })
+            },
+            (Some(_), None) => {
+                Err(D::Error::custom("`slots` is defined but no `values` field was found."))
```

Comment:
Nit: Error messages should be lowercase and not use trailing punctuation.

Applies to more messages in this file.

Reply:
Changed this (and while reviewing some code unrelated to this PR I fixed a couple that were not following the convention)